core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+use {
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        collections::{hash_map::Entry, HashMap},
+        fmt::{Debug, Display},
+        ops::{BitAndAssign, Sub},
+    },
+};
+
+pub const MAX_THREADS: usize = u64::BITS as usize;
+
+/// Identifier for a thread
+pub type ThreadId = usize; // 0..MAX_THREADS-1
+
+/// A bit-set of threads an account is scheduled or can be scheduled for.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ThreadSet(u64);
+
+/// Thread-aware account locks which allows for scheduling on threads
+/// that already hold locks on the account. This is useful for allowing
+/// queued transactions to be scheduled on a thread while the transaction
+/// is still being executed on the thread, up to a queue limit.
+pub struct ThreadAwareAccountLocks {
+    /// Number of threads.
+    num_threads: usize, // 0..MAX_THREADS
+    /// Limit on the number of sequentially-queued transactions per account.
+    sequential_queue_limit: u32,
+    /// Write locks - only on thread can hold a write lock at a time.
+    /// Contains how many write locks are held by the thread.
+    write_locks: HashMap<Pubkey, (ThreadId, u32)>,
+    /// Read locks - multiple threads can hold a read lock at a time.
+    /// Contains thread-set for easily checking which threads are scheduled.
```

Comment:
hmm, is there any actual perf. win due to `ThreadSet`? as far as i looked the code, this is used with `[u32; MAX_THREADS]`. and i couldn't spot any shortcut logic... could you provide some micro-bench maybe?

Reply:
... or just pin-point some piece of code which is specifically benefiting from `ThreadSet`.

Reply:
maybe, reading `[u32; MAX_THREADS]` repeatedly a lot just slows down _that much_?

Reply:
in `write_schedulable_threads`, we use the thread_set to quickly check `only_one_contained` instead of looping through the counts for each thread.